### PR TITLE
Implement bound_sympy

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102611
* #104558
* __->__ #104559
* #104557

The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumption was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

We implement a function that, given bounds on the values of the free
symbols of a sympy expression, it gives a bound on a the expression
itself.

We reshuffle a few things in value_ranges.py to create a
`SymPyValueRangeAnalysis` class, but we do not change any code really.
The only relevant change in that file is the addition of the
`sympy_bound`s function. We do this because we don't want to inadvertently
use any fallbacks in this case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

### CODE REVIEW

**Review by eellison — APPROVED** [review]
Please separate out refactorings from actual code changes, it makes it a lot easier for the reviewer and your PRs will get reviewed quicker.

**Comment by eellison** [comment]
this has test failures

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -504,3 +512,14 @@ def ceil(cls, x):
...
+
+def bound_sympy(expr: sympy.Expr, ranges: Dict[sympy.Symbol, ValueRanges]) -> ValueRanges:
+    # Add dynamic shapes within the expression as potentially unbounded
+    dynamic_shapes = expr.free_symbols - ranges.keys()
+    if dynamic_shapes:
+        ranges = deepcopy(ranges)
+        for s in dynamic_shapes:
+            ranges[s] = ValueRanges(0, math.inf)  # type: ignore[index]
+
+    return sympy_interp(SymPyValueRangeAnalysis(), ranges, expr)
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/utils/_sympy/value_ranges.py:525]: Yep, there's nothing else to review here.

**Comment by eellison** [comment]
Eventually, I think we would want to get add guards for what the symbolic expressions, instead of deferring all symbolic shapes to be unbounded. Planned at any point?

**Comment by lezcano** [comment]
>add guards for what the symbolic expressions

what do you mean by that exactly?

**Comment by eellison** [comment]
With optimize_indexing, for instance, eventually we'd want to guard so that we have two regimes, one where we can use int32 indexing and one with int64 indexing

**Comment by lezcano** [comment]
Yeah, that should certainly be doable.

**Pull request closed without merging**